builtin/providers/openstack/resource_openstack_fw_firewall_v2.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Firewall created: %#v", firewall)
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"PENDING_CREATE"},
+		Target:     "ACTIVE",
+		Refresh:    WaitForFirewallActive(networkingClient, firewall.ID),
+		Timeout:    30 * time.Second,
+		Delay:      0,
+		MinTimeout: 2 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+
+	d.SetId(firewall.ID)
+
+	return nil
+}
+
+func resourceFirewallRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] Retrieve information about firewall: %s", d.Id())
+
+	config := meta.(*Config)
+	networkingClient, err := config.networkingV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack networking client: %s", err)
+	}
+
+	firewall, err := firewalls.Get(networkingClient, d.Id()).Extract()
+	if err != nil {
+		httpError, ok := err.(*gophercloud.UnexpectedResponseCodeError)
+		if !ok {
+			return err
+		}
+
+		if httpError.Actual == 404 {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	if t, exists := d.GetOk("name"); exists && t != "" {
+		d.Set("name", firewall.Name)
+	} else {
+		d.Set("name", "")
+	}
+
+	if t, exists := d.GetOk("description"); exists && t != "" {
+		d.Set("description", firewall.Description)
+	} else {
+		d.Set("description", "")
+	}
+
+	if t, exists := d.GetOk("policy_id"); exists && t != "" {
+		d.Set("policy_id", firewall.PolicyID)
+	} else {
+		d.Set("policy_id", "")
+	}
+
+	if t, exists := d.GetOk("admin_state_up"); exists && t != "" {
+		d.Set("admin_state_up", firewall.AdminStateUp)
+	} else {
+		d.Set("admin_state_up", "")
+	}
+
+	if t, exists := d.GetOk("tenant_id"); exists && t != "" {
+		d.Set("tenant_id", firewall.TenantID)
+	} else {
+		d.Set("tenant_id", "")
+	}
+
+	return nil
+}
+
+func resourceFirewallUpdate(d *schema.ResourceData, meta interface{}) error {
+
+	config := meta.(*Config)
+	networkingClient, err := config.networkingV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack networking client: %s", err)
+	}
+
+	opts := firewalls.UpdateOpts{}
+
+	if d.HasChange("name") {
+		opts.Name = d.Get("name").(string)
+	}
+
+	if d.HasChange("description") {
+		opts.Description = d.Get("description").(string)
+	}
+
+	if d.HasChange("policy_id") {
+		opts.PolicyID = d.Get("policy_id").(string)
+	}
+
+	if d.HasChange("admin_state_up") {
+		adminStateUp := d.Get("admin_state_up").(bool)
+		opts.AdminStateUp = &adminStateUp
+	}
+
+	log.Printf("[DEBUG] Updating firewall with id %s: %#v", d.Id(), opts)
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"PENDING_CREATE", "PENDING_UPDATE"},
+		Target:     "ACTIVE",
+		Refresh:    WaitForFirewallActive(networkingClient, d.Id()),
+		Timeout:    30 * time.Second,
+		Delay:      0,
+		MinTimeout: 2 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+
+	return firewalls.Update(networkingClient, d.Id(), opts).Err
```

Comment:
Oh, I see. You're waiting for a pending create or update request to finish before calling this update request. Then perhaps just move the `Update` function one lineup and add the return statement above.